File: leetcode/design-browser-history.py
# 1472. Design Browser History
# 🟠 Medium
#
# https://leetcode.com/problems/design-browser-history/
#
# Tags: Array - Linked List - Stack - Design - Doubly-Linked List - Data Stream

import timeit


# We can optimize the previous solution and make all operations run in
# O(1) time if we use only one stack and keep both ints, one to know
# the index of the element that we are at currently, the other to
# determine how many steps forward we can move at this time.
#
# Time complexity: O(n) - Back and forward need to pop/push step
# elements between stacks, that is n where n is the value of steps.
# Space complexity: O(n) - Each call could result in one element added.
#
# Runtime 286 ms Beats 45.44%
# Memory 16.7 MB Beats 70.44%
class BrowserHistory:
    # O(1)
    def __init__(self, homepage: str):
        # Use two stacks to preserve back and forwards history.
        self.hist, self.f_hist = [homepage], []

    # O(1)
    def visit(self, url: str) -> None:
        # Push the url into the back stack and clear the forward stack.
        self.hist.append(url)
        if self.f_hist:
            self.f_hist = []

    # O(n) - Where n is the value of steps.
    def back(self, steps: int) -> str:
        while steps > 0 and len(self.hist) > 1:
            self.f_hist.append(self.hist.pop())
            steps -= 1
        return self.hist[-1]

    # O(n) - Where n is the value of steps.
    def forward(self, steps: int) -> str:
        while steps > 0 and self.f_hist:
            self.hist.append(self.f_hist.pop())
            steps -= 1
        return self.hist[-1]


# We can optimize the previous solution and make all operations run in
# O(1) time if we use only one stack and keep both ints, one to know
# the index of the element that we are at currently, the other to
# determine how many steps forward we can move at this time.
#
# Time complexity: O(1) - All operations are O(1).
# Space complexity: O(n) - Each call could result in one element added.
#
# Runtime 217 ms Beats 86.77%
# Memory 16.7 MB Beats 10.57%
class BrowserHistory2:
    def __init__(self, homepage: str):
        # Use a single stack and pointers.
        self.hist = [homepage]
        self.current_idx = 0
        self.can_move_forward = 0

    def visit(self, url: str) -> None:
        # Always move the current index and delete forward history.
        self.current_idx += 1
        self.can_move_forward = 0
        # If we are at the end of the history, append.
        if self.current_idx == len(self.hist):
            self.hist.append(url)
        else:
            self.hist[self.current_idx] = url

    def back(self, steps: int) -> str:
        # Compute the real number of steps we can take.
        real = min(steps, self.current_idx)
        self.current_idx -= real
        self.can_move_forward += real
        return self.hist[self.current_idx]

    def forward(self, steps: int) -> str:
        real = min(steps, self.can_move_forward)
        self.current_idx += real
        self.can_move_forward -= real
        return self.hist[self.current_idx]


def test():
    executors = [
        BrowserHistory,
        BrowserHistory2,
    ]
    tests = [
        [
            [
                "BrowserHistory",
                "visit",
                "visit",
                "visit",
                "back",
                "back",
                "forward",
                "visit",
                "forward",
                "back",
                "back",
            ],
            [
                ["leetcode.com"],
                ["google.com"],
                ["facebook.com"],
                ["youtube.com"],
                [1],
                [1],
                [1],
                ["linkedin.com"],
                [2],
                [2],
                [7],
            ],
            [
                None,
                None,
                None,
                None,
                "facebook.com",
                "google.com",
                "facebook.com",
                None,
                "linkedin.com",
                "google.com",
                "leetcode.com",
            ],
        ],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor(t[1][0][0])
                for i in range(1, len(t[0])):
                    call = t[0][i]
                    param = t[1][i][0]
                    result = getattr(sol, call)(t[1][i][0])
                    exp = t[2][i]
                    assert result == exp, (
                        f"\033[93m» {result} <> {exp}\033[91m for"
                        + f" test {col} using \033[1m{executor.__name__}"
                    )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/design-browser-history.rs
// 1472. Design Browser History
// 🟠 Medium
//
// https://leetcode.com/problems/design-browser-history/
//
// Tags: Array - Linked List - Stack - Design - Doubly-Linked List - Data Stream

struct BrowserHistory {
    hist: Vec<String>,
    current_idx: usize,
    can_move_forward: usize,
}

/**
* Use a single vector and two integers to make all the operations run in
* O(1), one integer marks the current index in the vector that we are
* located at currently, the other determines how many steps forward in
* the history we can move, if any.
*
* Time complexity: O(1) - All methods run in constant time.
* Space complexity: O(n) - Each call to visit will add one element to the
* vector.
*
* Runtime 38 ms Beats 54.55%
* Memory 4.6 Beats 63.64%
*/
impl BrowserHistory {
    fn new(homepage: String) -> Self {
        Self {
            hist: vec![homepage],
            current_idx: 0,
            can_move_forward: 0,
        }
    }

    fn visit(&mut self, url: String) {
        self.current_idx += 1;
        self.can_move_forward = 0;
        if self.current_idx == self.hist.len() {
            self.hist.push(url);
        } else {
            self.hist[self.current_idx] = url;
        }
    }

    fn back(&mut self, steps: i32) -> String {
        let real = (steps as usize).min(self.current_idx);
        self.current_idx -= real;
        self.can_move_forward += real;
        self.hist[self.current_idx].clone()
    }

    fn forward(&mut self, steps: i32) -> String {
        let real = (steps as usize).min(self.can_move_forward);
        self.current_idx += real;
        self.can_move_forward -= real;
        self.hist[self.current_idx].clone()
    }
}

// Tests.
fn main() {
    let mut bh = BrowserHistory::new(String::from("leetcode.com"));
    bh.visit(String::from("google.com"));
    bh.visit(String::from("facebook.com"));
    bh.visit(String::from("youtube.com"));
    assert_eq!(bh.back(1), String::from("facebook.com"));
    assert_eq!(bh.back(1), String::from("google.com"));
    assert_eq!(bh.forward(1), String::from("facebook.com"));
    bh.visit(String::from("linkedin.com"));
    assert_eq!(bh.forward(2), String::from("linkedin.com"));
    assert_eq!(bh.back(2), String::from("google.com"));
    assert_eq!(bh.back(7), String::from("leetcode.com"));
    println!("All tests passed!")
}
